allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+from typing import Callable, Dict, List, Union
+
+import torch
+
+from .common.checks import ConfigurationError
+from .common import Params
+
+
+class NlpApi:
+    """
+    The ``NlpApi`` provides a high-level API for building NLP models.  You don't have to use it
+    for building a model with AllenNLP, but we think you should, because it's useful.
+
+    ``NlpApi`` abstracts away particular decisions that are frequently necessary in NLP, like how
```

Comment:
A note on magic. Consider making the dataflow explicit. Instead of having nlp.get_recurrent_cell have some magic connection to the config file, maybe there is a way to do something like:
```
model.py
nlp.get_recurrent_cell(params.get("my_rnn_kind"))

experiment.json
"my_rnn_kind" = "bilstm"
```

or even:
```
model.py
nlp.get_recurrent_cell(params.get("my_rnn"))

experiment.json
"my_rnn": {
  "kind": "bilstm"
  "num_layers": 3
}
```